### PR TITLE
Fixes #5906 - Makes NaN respect the sign bit

### LINKED ISSUE #5906: The negative NaN's sign bit is inconsistent when running -NoNative
```
arr = new Float64Array([NaN]);
arr = new Uint8ClampedArray(arr.buffer);
// output:
// ch bug.js:              arr = [0,0,0,0,0,0,248,255]
// ch -NoNative bug.js:    arr = [0,0,0,0,0,0,248,255]


arr = new Float64Array([-NaN]);
arr = new Uint8ClampedArray(arr.buffer);
// output:
// ch bug.js:              arr = [0,0,0,0,0,0,248,255]
// ch -NoNative bug.js:    arr = [0,0,0,0,0,0,248,127]
```

NaN is represented by a float using the IEEE 754 floating point representation. When representing NaN, IEEE 754 uses the most significant bit as the sign bit. 

So NaN in IEEE 754 is represented by:
s111 1111 1111 xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
S = sign
xxx... = anything but all zeros

This explains the diff in outputs: new Uint8ClampedArray(arr.buffer); is taking the raw NaN value and putting it into an array where each index stores an 8 bit value (the transfer is done using Big-endian). The sign bit is apparent in the 7th index of arr: 255 means the sign bit is 1 (negative) and 127 means the sign bit is 0 (positive)

### PR DESCRIPTION
**Pull request by wyrichte**

### CODE REVIEW

File: lib/Common/Common/NumberUtilities.inl [hunk 1]
```diff
@@ -138,7 +138,8 @@ namespace Js
     {
 #if defined(TARGET_64)
         // NaN is a range of values; all bits on the exponent are 1's and some nonzero significant.
-        // no distinction on signed NaN's
+        // No distinction on signed NaN's unless requirePositive is true, then true is returned only
+        // when value is NaN and the sign bit of value is zero.
```
Comment (rajatd) [inline lib/Common/Common/NumberUtilities.inl]: I don't see code corresponding to this comment.

File: lib/Runtime/Library/JavascriptNumber.cpp [hunk 2]
```diff
@@ -23,7 +23,7 @@ namespace Js
 #if FLOATVAR
         if (IsNan(value))
         {
-            value = JavascriptNumber::NaN;
+            value = IsNegative(value) ? JavascriptNumber::NegativeNaN : JavascriptNumber::NaN;
```
Comment (rajatd) [inline lib/Runtime/Library/JavascriptNumber.cpp]: >IsNegative [](start = 20, length = 10)

There might be other places where we canonicalize NaNs.
For example, `JavascriptConversion::TryCanonicalizeAsSimpleVar`.

**Comment by rajatd** [comment]
There isn't a difference in behavior between interpreter and jit, is there?

**Comment by MikeHolman** [comment]
What are we trying to accomplish with this?

As far as I know, the only time the bit-pattern/sign of NaN is observable is when storing/reading from a typed array (and spec doesn't mandate anything it). When reading, we already need to canonicalize. We could canonicalize before stores if we are trying to make it totally opaque (but there may be a cost there).

**Comment by wyrichte** [comment]
> What are we trying to accomplish with this?

[Fixing this bug](https://github.com/Microsoft/ChakraCore/issues/5906
)
> 
> and spec doesn't mandate anything it

Agreed. The goal of this commit is to make sure Chakra is consistent (i.e. make sure the same array buffer contents are seen independent of the -NoNative flag). The spec does not mandate the behavior of the sign bit of NaN.

File: lib/Runtime/Library/JavascriptNumber.inl [hunk 3]
```diff
@@ -148,7 +148,7 @@ namespace Js
     inline Var JavascriptNumber::ToVar(double value)
     {
         uint64 val = *(uint64*)&value;
-        AssertMsg(!IsNan(value) || ToSpecial(value) == k_Nan || ToSpecial(value) == 0x7FF8000000000000ull, "We should only produce a NaN with this value");
+//        AssertMsg(!IsNan(value) || ToSpecial(value) == k_Nan || ToSpecial(value) == 0x7FF8000000000000ull, "We should only produce a NaN with this value");
```
Comment (LouisLaf) [inline lib/Runtime/Library/JavascriptNumber.inl]: >0x7FF8000000000000ull [](start = 86, length = 21)

Shouldn't this just be k_NegativeNan now?

File: lib/Runtime/Library/JavascriptNumber.cpp (hunk 2 above)
Comment (LouisLaf) [reply lib/Runtime/Library/JavascriptNumber.cpp]: As well as JavascriptNumber::TryToVarFastWithCheck() and JavascriptNumber::NewWithCheck()

---
In reply to: [246601736](https://github.com/Microsoft/ChakraCore/pull/5905#discussion_r246601736) [](ancestors = 246601736)

File: lib/Runtime/Library/JavascriptNumber.inl (hunk 3 above)
Comment (wyrichte) [reply lib/Runtime/Library/JavascriptNumber.inl]: Yes

**Pull request merged**